File: src/models/tokenizer/NgramSearchTokenizer.java
package models.tokenizer;

import java.io.File;
import java.util.*;

public class NgramSearchTokenizer implements Tokenizer {
    private static int MIN = 3;
    private static int MAX = 30;

    @Override
    public void tokenizeData(File file, String text, Map<String, Map<File, Integer>> data) {

        Arrays.stream(text.split(SPLITTER)).forEach(token -> {
            if (token.length() > MIN) {
                final int max = Math.min(MAX, token.length() - 1);

                for (int length = MIN; length <= max; length++) {
                    for (int start = 0; start + length <= token.length(); start++) {
                        String newToken = token.substring(start, start + length);

                        Map<File, Integer> occurrences = data.get(newToken);
                        if (occurrences != null) {
                            Integer occurrence = occurrences.get(file);
                            if (occurrence != null) {
                                occurrences.replace(file, occurrence + 1);
                            } else {
                                occurrences.put(file, 1);
                            }
                        } else {
                            Map<File, Integer> occurrence = new HashMap<>();
                            occurrence.put(file, 1);
                            data.put(newToken, occurrence);
                        }
                    }
                }
            }
        });
    }

    @Override
    public Set<String> tokenizeQuery(String text) {
        Set<String> tokenSet = new HashSet<>();

        Arrays.stream(text.split(SPLITTER)).forEach(token -> {
            if (token.length() > MIN) {
                final int max = Math.min(MAX, token.length() - 1);

                for (int length = MIN; length <= max; length++) {
                    for (int start = 0; start + length <= token.length(); start++) {
                        tokenSet.add(token.substring(start, start + length));
                    }
                }
            }
        });

        return tokenSet;
    }

    @Override
    public LinkedList<String> develop(String token) {
        return new LinkedList<>(Collections.singletonList(token));
    }
}

File: C#/SimpleSearch/models/tokenizer/NgramSearchTokenizer.cs
using SimpleSearch.models.search;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text.RegularExpressions;

namespace SimpleSearch.models.tokenizer
{
    class NgramSearchTokenizer : Tokenizer
    {
        private static readonly int MIN = 3;
        private static readonly int MAX = 30;

        public override ISet<string> TokenizeQuery(string text)
        {
            ISet<string> tokenSet = new HashSet<string>();

            Regex.Split(text, SPLITTER).ToList().ForEach(token =>
            {
                if (token.Length > MIN)
                {
                    int max = Math.Min(MAX, token.Length);

                    for (int length = MIN; length <= MAX; length++)
                    {
                        for (int start = 0; start + length < +token.Length; start++)
                        {
                            tokenSet.Add(token.Substring(start, length));
                        }
                    }
                }
            });

            return tokenSet;
        }
        public override List<string> Develope(string token) => (new string[] { token }).ToList();

        public override void tokenizeData(string filePath, string text, IDictionary<string, InvertedIndex> data)
        {
            Regex.Split(text, SPLITTER).ToList().ForEach(token =>
            {
                if (token.Length > MIN)
                {
                    int max = Math.Min(MAX, token.Length);

                    for (int length = MIN; length <= MAX; length++)
                    {
                        for (int start = 0; start + length < token.Length; start++)
                        {
                            string newToken = token.Substring(start, length);

                            if (data.ContainsKey(newToken))
                            {
                                if (data[newToken].ContainsKey(filePath))
                                {
                                    data[newToken][filePath]++;
                                }
                                else
                                {
                                    data[newToken][filePath] = 1;
                                }
                            }
                            else
                            {
                                data[newToken] = new InvertedIndex
                                {
                                    [filePath] = 1
                                };
                            }
                        }
                    }
                }
            });
        }
    }
}
